File: packages/ts/src/engine.ts
import { lookupRule, MANIFEST } from "./manifest.js";
import type { Ctx, Finding, ResolvedOptions } from "./types.js";
import { RULES } from "./rules/index.js";

export interface Analyzed {
  idArr: string[] | null;
  sizeArr: number[] | null;
  idSet: Set<string> | null;
  product: bigint | null;
  productOverflow: boolean;
}

/** Compute id/size arrays, the valid id set, and the safe (overflow-checked) product(size). */
export function analyzeDataset(doc: any, maxCells: number): Analyzed {
  const rawId = Array.isArray(doc?.id) ? (doc.id as any[]) : null;
  const rawSize = Array.isArray(doc?.size) ? (doc.size as any[]) : null;
  const idArr = rawId as string[] | null;
  const idSet = rawId ? new Set(rawId.filter((x: any) => typeof x === "string") as string[]) : null;

  let product: bigint | null = null;
  let productOverflow = false;
  if (rawSize && rawSize.every((s: any) => typeof s === "number" && Number.isInteger(s) && s >= 0)) {
    const sizes = rawSize as number[];
    let p = 1n;
    for (const s of sizes) {
      p *= BigInt(s);
      if (p > BigInt(maxCells)) {
        productOverflow = true;
        break;
      }
    }
    product = productOverflow ? null : p;
  }
  return { idArr, sizeArr: rawSize as number[] | null, idSet, product, productOverflow };
}

/** Run all registered rules against a dataset node, pushing findings into `sink`. */
export function runSemantic(
  doc: any,
  rootPointer: string,
  opts: ResolvedOptions,
  analyzed: Analyzed,
  sink: (f: Finding) => void,
): void {
  const finding: Ctx["finding"] = (ruleId, partial) => {
    const r = lookupRule(ruleId);
    sink({
      code: r.code,
      ruleId: r.id,
      severity: r.severity,
      specRef: r.specRef,
      ...partial,
    });
  };

  const ctx: Ctx = {
    doc,
    rootPointer,
    opts,
    manifest: MANIFEST,
    finding,
    product: analyzed.product,
    productOverflow: analyzed.productOverflow,
    idArr: analyzed.idArr,
    sizeArr: analyzed.sizeArr,
    idSet: analyzed.idSet,
  };

  const dims = doc?.dimension && typeof doc.dimension === "object" ? (doc.dimension as Record<string, any>) : null;

  for (const rule of RULES) {
    if (rule.kind === "dataset") {
      rule.check(ctx);
    } else {
      if (!dims || !analyzed.idArr) continue; // degraded: cannot map dimensions to sizes
      for (let i = 0; i < analyzed.idArr.length; i++) {
        const dimId = analyzed.idArr[i];
        if (!dimId || typeof dimId !== "string") continue;
        const dim = dims[dimId];
        if (!dim || typeof dim !== "object") continue;
        const size = analyzed.sizeArr && i < analyzed.sizeArr.length ? (analyzed.sizeArr[i] ?? null) : null;
        rule.check(ctx, dimId, dim, size);
      }
    }
  }
}

File: crates/validator/src/engine.rs
use crate::manifest::ManifestData;
use crate::rules;
use crate::types::{Finding, ValidateOptions};
use serde_json::Value;
use std::collections::HashSet;

pub struct Analyzed {
    pub id_arr: Option<Vec<String>>,
    pub size_arr: Option<Vec<i64>>,
    pub id_set: Option<HashSet<String>>,
    pub product: Option<u128>,
    pub overflow: bool,
}

/// Compute id/size arrays, the valid id set, and the safe (overflow-checked) product(size).
pub fn analyze_dataset(doc: &Value, max_cells: u64) -> Analyzed {
    let id_arr = doc["id"]
        .as_array()
        .map(|a| a.iter().filter_map(|v| v.as_str().map(|s| s.to_string())).collect::<Vec<_>>());

    let id_set = id_arr.as_ref().map(|v| v.iter().cloned().collect::<HashSet<_>>());

    let size_arr = doc["size"].as_array().and_then(|a| {
        let mut out = Vec::with_capacity(a.len());
        for x in a {
            match x.as_i64() {
                Some(n) if n >= 0 => out.push(n),
                _ => return None,
            }
        }
        Some(out)
    });

    let mut overflow = false;
    let mut product: Option<u128> = None;
    if let Some(ref sizes) = size_arr {
        let mut p: u128 = 1;
        for &s in sizes {
            match p.checked_mul(s as u128) {
                Some(x) => p = x,
                None => {
                    overflow = true;
                    break;
                }
            }
            if p > max_cells as u128 {
                overflow = true;
                break;
            }
        }
        if !overflow {
            product = Some(p);
        }
    }

    Analyzed { id_arr, size_arr, id_set, product, overflow }
}

pub struct Ctx<'a> {
    pub doc: &'a Value,
    pub root: &'a str,
    pub opts: &'a ValidateOptions,
    pub mf: &'a ManifestData,
    pub analyzed: &'a Analyzed,
}

/// Run all registered rules against a dataset node, appending findings into `out`.
pub fn run_semantic(
    doc: &Value,
    root: &str,
    opts: &ValidateOptions,
    mf: &ManifestData,
    analyzed: &Analyzed,
    out: &mut Vec<Finding>,
) {
    let ctx = Ctx { doc, root, opts, mf, analyzed };
    rules::dataset_rules(&ctx, out);
    rules::dimension_rules(&ctx, out);
}
